libs/crypto/merkle.cpp

```diff
@@ -0,0 +1,273 @@
+/*
+ * Returns whether or not init has been called.
+ */
+bool Merkle::isInitialized() {
+  return initialized;
+}
+
+/*
+ * Returns whether or not the tree has finished building.
+ * The variable buildingState counts down from the number of nodes
+ * needed in the tree. When it reaches 0, all nodes have been
+ * created.
+ */
+bool Merkle::isCompleted() {
+  return (buildingState == 0);
+}
+
+/*
+ * Calls the operation method until the tree has been built.
+ */
+void Merkle::buildTree() {
+  while(buildingState > 0) {
+    operation();
+  }
+}
+
+/*
+ * Does one "operation" to build new tree.
+ * An operation makes one node, either a leaf node or a normal
+ * node.
+ */
+void Merkle::operation(){
+  if (buildingState <= 0) return;
+  if (needLeaf()) {
+    makeLeaf();
+  } else {
+    combine();
+  }
+  buildingState--;
+  if (buildingState <= 0) msg = 0;
+}
+
+/*
+ * Determines if the next node to create is a leaf node
+ * using the stack, s, of nodes.
+ */
+bool Merkle::needLeaf(){
+  if (s.empty()) return true;
+  Node x = s.top();
+  s.pop();
+  if (s.empty()){
+    s.push(x);
+    return true;
+  }
+  Node y = s.top();
+  s.push(x);
+  return (x.height != y.height);
+}
+
+/*
+ * Makes a new leaf to put in the leaves of the tree.
+ * Otherwise, it's just pushed onto the stack for continued calculation
+ */
+void Merkle::makeLeaf() {
+  Data sk = Data::generateSecretKey(secretKey, msg, digestsize);
+  msg++;
+  Winternitz sig(sk, ell);
+  Node leaf;
+  leaf.data = sig.getPublicKey();
+  leaf.height = 0;
+  s.push(leaf);
+  winter.push_back(sig); // add leaf to winter
+  newEntry(leaf);
+}
+
+/*
+ * Combines top two entries in stack -- assumes they are at correct height.
+ * Makes a new node at the height + 1
+ */
+void Merkle::combine(){
+  vector<Data> vec;
+  vec.push_back(s.top().data); s.pop();
+  unsigned int height = s.top().height;
+  vec.push_back(s.top().data); s.pop();
+  Data combine = Data::combineHashes(vec, DIGESTSIZE);
+  Node c = {combine, height+1};
+  newEntry(c);
+  s.push(c);
+}
+
+/*
+ * Enters a new node into the next spot in the tree at the corresponding
+ * height
+ * If height not there, adds enough to get there :)
+ */
+void Merkle::newEntry(Node entry) {
+  while (tree.size() < entry.height + 1) {
+    tree.push_back(vector<Data>() );
+  }
+  tree[entry.height].push_back(entry.data);
+}
+
+
+/*** Signing and Verifying ***/
+
+/*
+ * Signs a message digest using the next leaf node in the tree, if
+ * one is available. Does not do error handling.
+ */
+Merkle::Signature Merkle::sign(Data digest) {
+  // This is just a bitshift -- O(1), no need to store the value
+  if (msg >= (1 << depth)) {
+    messagesException e;
+    throw e;
+  }
+  long msgNum = getNextMsg();
+  Signature merkSig;
+  merkSig.wint = winter[msgNum].sign(digest);
+  merkSig.auth = getAuthPath(msgNum);
+  merkSig.msgNum = msgNum;
+  return merkSig;
+}
+
+/*
+ * In the subclass, this is a critical region.
+ */
+long Merkle::getNextMsg() {
+  msg++;
```

Comment:
quite bad to have side-effects on functions with get-names

Reply:
Thanks @l4l for pointing this out. But any alternative for get-names?

Reply:
Maybe just `nextMsg` ?